spikefeatures/features.py

```diff
@@ -1 +1,203 @@
-# port waveform_features_utils.py here
\ No newline at end of file
+"""
+Functions based on
+https://github.com/AllenInstitute/ecephys_spike_sorting/blob/master/ecephys_spike_sorting/modules/mean_waveforms/waveform_metrics.py
+22/04/2020
+"""
+
+import numpy as np
+from scipy.stats import linregress
+import pandas as pd
+
+all_1D_features = ['peak_to_valley', 'halfwidth', 'peak_trough_ratio',
+                   'repolarization_slope', 'recovery_slope']
+
+
+def calculate_features(waveforms, sampling_frequency, feature_names=None,
+                       recovery_slope_window=0.7):
+    """ Calculate features for all waveforms
+
+    Parameters
+    ----------
+    waveforms  : numpy.ndarray (num_waveforms x num_samples)
+        waveforms to compute features for
+    sampling_frequency  : float
+        rate at which the waveforms are sampled (Hz)
+    feature_names : list or None (if None, compute all)
+        features to compute
+    recovery_slope_window : float
+        windowlength in ms after peak wherein recovery slope is computed
+
+    Returns
+    -------
+    metrics : pandas.DataFrame  (num_waveforms x num_metrics)
+        one column for each metric
+        one row per waveforms
+
+    """
+    metrics = pd.DataFrame()
+
+    if feature_names is None:
+        feature_names = all_1D_features
+    else:
+        for name in feature_names:
+            assert name in all_1D_features, f'{name} not in {all_1D_features}'
+
+    if 'peak_to_valley' in feature_names:
+        metrics['peak_to_valley'] = peak_to_valley(waveforms=waveforms,
+                                                   sampling_frequency=sampling_frequency)
+    if 'peak_trough_ratio' in feature_names:
+        metrics['peak_trough_ratio'] = peak_trough_ratio(waveforms=waveforms)
+
+    if 'halfwidth' in feature_names:
+        metrics['halfwidth'] = halfwidth(waveforms=waveforms,
+                                         sampling_frequency=sampling_frequency)
+
+    if 'repolarization_slope' in feature_names:
+        metrics['repolarization_slope'] = repolarization_slope(
+            waveforms=waveforms,
+            sampling_frequency=sampling_frequency,
+        )
+
+    if 'recovery_slope' in feature_names:
+        metrics['recovery_slope'] = recovery_slope(
+            waveforms=waveforms,
+            sampling_frequency=sampling_frequency,
+            window=recovery_slope_window,
+        )
+    return metrics
+
+
+def peak_to_valley(waveforms, sampling_frequency):
+    """ time in s between throug and peak
+
+    :param waveforms: numpy.ndarray (num_waveforms x num_samples)
+    :param sampling_frequency: rate at which the waveforms are sampled (Hz)
+    :return: peak_to_valley; np.ndarray (num_waveforms)
+    """
+    trough_idx, peak_idx = _get_trough_and_peak_idx(waveforms)
+    ptv = (peak_idx - trough_idx) * (1/sampling_frequency)
+    return ptv
+
+
+def peak_trough_ratio(waveforms):
+    """ waveform peak height / trough depth
```

Comment:
@TRuikes one last thing: can you use the same notation for the docstrings?

```
Parameters
    ----------
    waveforms  : numpy.ndarray (num_waveforms x num_samples)
        waveforms to compute features for
    sampling_frequency  : float
        rate at which the waveforms are sampled (Hz)
    feature_names : list or None (if None, compute all)
        features to compute
    recovery_slope_window : float
        windowlength in ms after peak wherein recovery slope is computed
    Returns
    -------
    metrics : pandas.DataFrame  (num_waveforms x num_metrics)
        one column for each metric
        one row per waveforms
```

instead of: 

```
""" Width of the waveform peak at its half ampltude heigth
    :param return_idx: bool, if true return halfwidth, index of crossing threhold pre peak, index of crossing post peak
    :param waveforms: numpy.ndarray (num_waveforms x num_samples)
    :param sampling_frequency: rate at which the waveforms are sampled (Hz)
    :return: peak_to_valley; np.ndarray (num_waveforms)
    """
```

Reply:
ok will do